biogtr/models/gtr_runner.py

```diff
@@ -0,0 +1,191 @@
+            with length `clip_length` where each dict is a frame
+            batch_idx: the batch number used by lightning
+
+        Returns:
+            A dict containing the val loss plus any other metrics specified
+        """
+        result = self._shared_eval_step(val_batch[0], eval_metrics=self.val_metrics)
+        for metric, val in result.items():
+            self.log(f"val_{metric}", val, batch_size=len(val_batch[0]))
+        return result
+
+    def test_step(self, test_batch: list[dict], batch_idx: int) -> dict[str, float]:
+        """Method outlining the test procedure for model.
+
+        Args:
+            val_batch: A single batch from the dataset which is a list of dicts
+            with length `clip_length` where each dict is a frame
+            batch_idx: the batch number used by lightning
+
+        Returns:
+            A dict containing the val loss plus any other metrics specified
+        """
+        result = self._shared_eval_step(test_batch[0], eval_metrics=self.test_metrics)
+        for metric, val in result.items():
+            self.log(f"val_{metric}", val, batch_size=len(test_batch[0]))
+        return result
+
+    def predict_step(self, batch: list[dict], batch_idx: int) -> dict:
+        """Method describing inference for model.
+
+        Computes association + assignment.
+
+        Args:
+            batch: A single batch from the dataset which is a list of dicts
+            with length `clip_length` where each dict is a frame
+            batch_idx: the batch number used by lightning
+
+        Returns:
+            A list of dicts where each dict is a frame containing the predicted track ids
+        """
+        tracker = Tracker(self.model, **self.tracker_cfg)
+        instances_pred = tracker(batch[0])
+        return instances_pred
+
+    def _shared_eval_step(self, instances, eval_metrics=["sw_cnt"]):
+        """Helper function for running evaluation used by train, test, and val steps.
+
+        Args:
+            instances: A list of dicts where each dict is a frame containing gt data
+            eval_metrics: A list of metrics calculated and saved
+
+        Returns:
+            a dict containing the loss and any other metrics specified by `eval_metrics`
+        """
+        if self.model.transformer.return_embedding:
+            logits, _ = self(instances)
+        else:
+            logits = self(instances)
+        loss = self.loss(logits, instances)
+
+        return_metrics = {"loss": loss}
+        if "sw_cnt" in eval_metrics:
+            tracker = Tracker(self.model, **self.tracker_cfg)
+            instances_pred = tracker(instances)
+            matches, indices, _ = metrics.get_matches(instances_pred)
+            switches = metrics.get_switches(matches, indices)
+            sw_cnt = metrics.get_switch_count(switches)
+            return_metrics["sw_cnt"] = sw_cnt
+        return return_metrics
+
+    def configure_optimizers(self) -> dict:
```

Comment:
Add functions to init optimizers and schedulers (i think this is added in subsequent PRs, please confirm).

Reply:
Yep see [model_utils](https://github.com/talmolab/biogtr/blob/d4ff53d5d237f3c183c3e4341617c2523fbb2c99/biogtr/models/model_utils.py#L71)